File: src/web/magmac/app/lang/MutableLazyRule.ts
export class MutableLazyRule {
	public lex( input : String) : CompileResult<Node> {return this.maybeRule.map( 0).orElseGet( ( )->CompileErrors.createStringError( "Rule not set", input));;}
	public generate( node : Node) : CompileResult<String> {return this.maybeRule.map( 0).orElseGet( ( )->CompileErrors.createNodeError( "Rule not set", node));;}
	public set( rule : Rule) : LazyRule {this.maybeRule=new Some<>( rule);return this;;}
}

File: src/java/magmac/app/lang/MutableLazyRule.java
package magmac.app.lang;

import magmac.api.None;
import magmac.api.Option;
import magmac.api.Some;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.error.error.CompileErrors;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.Rule;

public class MutableLazyRule implements LazyRule {
    private Option<Rule> maybeRule = new None<>();

    @Override
    public CompileResult<Node> lex(String input) {
        return this.maybeRule
                .map((Rule rule) -> rule.lex(input))
                .orElseGet(() -> CompileErrors.createStringError("Rule not set", input));
    }

    @Override
    public CompileResult<String> generate(Node node) {
        return this.maybeRule
                .map((Rule rule) -> rule.generate(node))
                .orElseGet(() -> CompileErrors.createNodeError("Rule not set", node));
    }

    @Override
    public LazyRule set(Rule rule) {
        this.maybeRule = new Some<>(rule);
        return this;
    }
}
